contrib/google.golang.org/api/internal/gen_endpoints/main.go

```diff
@@ -0,0 +1,270 @@
+		BaseURL       string                  `json:"baseUrl"`
+		BasePath      string                  `json:"basePath"`
+		Resources     map[string]*APIResource `json:"resources"`
+		RootURL       string                  `json:"rootUrl"`
+	}
+	APIResource struct {
+		Methods   map[string]*APIMethod   `json:"methods"`
+		Resources map[string]*APIResource `json:"resources"`
+	}
+	APIMethod struct {
+		ID         string `json:"id"`
+		FlatPath   string `json:"flatPath"`
+		Path       string `json:"path"`
+		HTTPMethod string `json:"httpMethod"`
+	}
+	Endpoint struct {
+		Hostname     string
+		HTTPMethod   string
+		PathTemplate string
+		PathMatcher  *regexp.Regexp
+
+		ServiceName  string
+		ResourceName string
+	}
+)
+
+// String returns a constructor without field names.
+func (e Endpoint) String() string {
+	return fmt.Sprintf(`{Hostname: "%s", HTTPMethod: "%s", PathTemplate: "%s", PathMatcher: regexp.MustCompile(`+"`"+`%s`+"`"+`), ServiceName: "%s", ResourceName: "%s"}`,
+		e.Hostname, e.HTTPMethod, e.PathTemplate, e.PathMatcher.String(), e.ServiceName, e.ResourceName)
+}
+
+func main() {
+	var es []Endpoint
+
+	root := downloadGoogleApiSrc()
+	defer os.RemoveAll(root)
+
+	log.Println("Parsing GCP service json API definitions...")
+
+	err := filepath.Walk(root, func(p string, fi os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if fi.IsDir() {
+			return nil
+		}
+		if filepath.Ext(p) == ".json" {
+			var def APIDefinition
+			f, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+
+			if err := json.NewDecoder(f).Decode(&def); err != nil {
+				return err
+			}
+			for _, resource := range def.Resources {
+				res, err := handleResource(&def, resource)
+				if err != nil {
+					return err
+				}
+				es = append(es, res...)
+			}
+		}
+
+		return nil
+	})
+	checkError(err)
+
+	f, err := os.Create(outFile)
+	checkError(err)
+	defer f.Close()
+
+	sort.Slice(es, func(i, j int) bool {
+		return es[i].String() < es[j].String()
+	})
+
+	log.Printf("Generating templated Go code in %s...\n", outFile)
+	template.Must(template.New("").Parse(tpl)).Execute(f, map[string]interface{}{
+		"Endpoints": es,
+		"Year":      time.Now().Year(),
+	})
+	log.Println("Done!")
+}
+
+func checkError(err error) {
+	if err != nil {
+		log.Fatalln(err)
+	}
+}
+
+func handleResource(def *APIDefinition, resource *APIResource) ([]Endpoint, error) {
+	var es []Endpoint
+	if resource.Methods != nil {
+		for _, method := range resource.Methods {
+			mes, err := handleMethod(def, method)
+			if err != nil {
+				return nil, err
+			}
+			es = append(es, mes...)
+		}
+	}
+	if resource.Resources != nil {
+		for _, child := range resource.Resources {
+			res, err := handleResource(def, child)
+			if err != nil {
+				return nil, err
+			}
+			es = append(es, res...)
+		}
+	}
+	return es, nil
+}
+
+func handleMethod(def *APIDefinition, method *APIMethod) ([]Endpoint, error) {
```

Comment:
It looks like this can only ever return a single endpoint?

Reply:
changed signature and renamed the method 👍